lib/Doctrine/DBAL/Types/TypeRegistry.php

```diff
@@ -0,0 +1,114 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Doctrine\DBAL\Types;
+
+use Doctrine\DBAL\DBALException;
+use function array_search;
+use function in_array;
+
+/**
+ * @internal
+ */
+final class TypeRegistry
+{
+    /** @var Type[] Map of type names and their corresponding objects. */
+    private $instances = [];
+
+    /**
+     * Factory method to create type instances.
+     * Type instances are implemented as flyweights.
+     *
+     * @throws DBALException
+     */
+    public function get(string $name) : Type
+    {
+        if (! isset($this->instances[$name])) {
+            throw DBALException::unknownColumnType($name);
+        }
+
+        return $this->instances[$name];
+    }
+
+    /**
+     * Find a name for the given type.
+     *
+     * @throws DBALException
+     */
+    public function lookupName(Type $type) : string
+    {
+        $name = $this->findTypeName($type);
+
+        if ($name === null) {
+            throw DBALException::typeNotRegistered($type);
+        }
+
+        return $name;
+    }
+
+    /**
+     * Checks if there is a type of the given name.
+     */
+    public function has(string $name) : bool
+    {
+        return isset($this->instances[$name]);
+    }
+
+    /**
+     * Adds a custom type to the type map.
+     *
+     * @throws DBALException
+     */
+    public function register(string $name, Type $type) : void
+    {
+        if (isset($this->instances[$name])) {
+            throw DBALException::typeExists($name);
+        }
+
+        if ($this->findTypeName($type) !== null) {
+            throw DBALException::typeAlreadyRegistered($type);
+        }
+
+        $this->instances[$name] = $type;
+    }
+
+    /**
+     * Overrides an already defined type to use a different implementation.
+     *
+     * @throws DBALException
+     */
+    public function override(string $name, Type $type) : void
+    {
+        if (! isset($this->instances[$name])) {
+            throw DBALException::typeNotFound($name);
+        }
+
+        if (! in_array($this->findTypeName($type), [$name, null], true)) {
```

Comment:
why don't you throw if the type is already registered to the same name ? Unless you consider that doing it again is harmless (it is, but for consistency check maybe it'd be better to throw imo)

Reply:
> Unless you consider that doing it again is harmless

Exactly for this reason, and also because the method is named _override_, so overriding by _itself_ doesn't seem harmful (unlike setting the same in _register_).